File: problems/equivalent-resistance/languages/python/resistor_utils.py
def series(a, b):
    return a + b


def parallel(a, b):
    return 1 / (1 / a + 1 / b)


def base_scf(index):
    return str(index)


def combine_scf(left, right, op):
    return f"({left}){op}({right})"


def evaluate_config(configuration, base_resistances):
    if configuration[0] != "(":
        return base_resistances[int(configuration)]

    parentheses = 1
    for i in range(1, len(configuration)):
        if configuration[i] == "(":
            parentheses += 1
        elif configuration[i] == ")":
            parentheses -= 1
        if parentheses == 0:
            start, end = _get_splits(configuration[i:])
            if start == -1:
                return evaluate_config(configuration[1:i], base_resistances)
            op = configuration[i + start : i + end]
            left = evaluate_config(configuration[1:i], base_resistances)
            right = evaluate_config(
                configuration[i + end + 1 : len(configuration) - 1],
                base_resistances,
            )
            return _FUNCTIONS[op](left, right)

    return -1


def _get_splits(config):
    op_symbols = ["+", "//"]
    for i in range(len(config)):
        for op in op_symbols:
            if config[i] == op[0]:
                return i, i + len(op)
    return -1, -1


_FUNCTIONS = {
    "+": lambda a, b: a + b,
    "//": lambda a, b: 1 / (1 / a + 1 / b),
}

File: problems/equivalent-resistance/languages/java/src/main/java/com/stephenacomb/ResistorUtils.java
package com.stephenacomb;

import java.util.Map;
import java.util.function.BiFunction;

public class ResistorUtils {

	public static double series(double a, double b) {
		return a + b;
	}

	public static double parallel(double a, double b) {
		return 1 / ((1 / a) + (1 / b));
	}

	public static String baseScf(int index) {
		return String.valueOf(index);
	}

	public static String combineScf(String left, String right, String op) {
		return "(" + left + ")" + op + "(" + right + ")";
	}

	public static double evaluateConfig(String configuration, double[] baseResistances) {
		int len = configuration.length();
		if (configuration.charAt(0) != '(')
			return baseResistances[Integer.parseInt(configuration)];
		else {
			int parentheses = 1;
			for (int i = 1; i < len; i++) {
				if (configuration.charAt(i) == '(')
					parentheses++;
				else if (configuration.charAt(i) == ')')
					parentheses--;
				if (parentheses == 0) {
					int[] splits = getSplits(configuration.substring(i));
					if (splits[0] == -1)
						return evaluateConfig(configuration.substring(1, i), baseResistances);
					BiFunction<Double, Double, Double> operation = functions().get(
							configuration.substring(i + splits[0], i + splits[1]));
					return operation.apply(
							Double.valueOf(
									evaluateConfig(
											configuration.substring(1, i), baseResistances)),
							Double.valueOf(
									evaluateConfig(
											configuration.substring(i + splits[1] + 1, len - 1), baseResistances)));
				}
			}
		}
		return -1;
	}

	private static int[] getSplits(String config) {
		int[] splits = new int[2];
		int firstOpChar = -1;
		int lastOpCharPP = -1;
		String[] opSymbols = new String[] { "+", "//" };
		for (int i = 0; i < config.length(); i++) {
			for (String opStr : opSymbols) {
				if (config.charAt(i) == opStr.charAt(0)) {
					firstOpChar = i;
					lastOpCharPP = i + opStr.length();
					splits[0] = firstOpChar;
					splits[1] = lastOpCharPP;
					return splits;
				}
			}
		}
		splits[0] = firstOpChar;
		splits[1] = lastOpCharPP;
		return splits;
	}

	private static Map<String, BiFunction<Double, Double, Double>> functions() {
		return Map.of(
				"+", (a, b) -> a + b,
				"//", (a, b) -> 1 / ((1 / a) + (1 / b)));
	}
}
